x/cdp/keeper/querier.go

```diff
@@ -178,3 +183,159 @@ func queryGetAccounts(ctx sdk.Context, req abci.RequestQuery, keeper Keeper) ([]
+	var params types.QueryCdpsParams
+	err := types.ModuleCdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrJSONUnmarshal, err.Error())
+	}
+
+	// Filter CDPs
+	filteredCDPs := FilterCDPs(ctx, keeper, params)
+
+	bz, err := codec.MarshalJSONIndent(keeper.cdc, filteredCDPs)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrJSONMarshal, err.Error())
+	}
+
+	return bz, nil
+}
+
+// FilterCDPs queries the store for all CDPs that match query params
+func FilterCDPs(ctx sdk.Context, k Keeper, params types.QueryCdpsParams) types.AugmentedCDPs {
+	var matchCollateralType, matchOwner, matchID, matchRatio types.CDPs
+
+	// match cdp owner (if supplied)
+	if len(params.Owner) > 0 {
+		denoms := k.GetCollateralTypes(ctx)
+		for _, denom := range denoms {
+			cdp, found := k.GetCdpByOwnerAndCollateralType(ctx, params.Owner, denom)
+			if found {
+				matchOwner = append(matchOwner, cdp)
+			}
+		}
+	}
+
+	// match cdp collateral denom (if supplied)
+	if len(params.CollateralType) > 0 {
+		// if owner is specified only iterate over already matched cdps for efficiency
+		if len(params.Owner) > 0 {
+			for _, cdp := range matchOwner {
+				if cdp.Type == params.CollateralType {
+					matchCollateralType = append(matchCollateralType, cdp)
+				}
+			}
+		} else {
+			matchCollateralType = k.GetAllCdpsByCollateralType(ctx, params.CollateralType)
+		}
+	}
+
+	// match cdp ID (if supplied)
+	if params.ID != 0 {
+		denoms := k.GetCollateralTypes(ctx)
+		for _, denom := range denoms {
+			cdp, found := k.GetCDP(ctx, denom, params.ID)
+			if found {
+				matchID = append(matchID, cdp)
+			}
+		}
+	}
+
+	// match cdp ratio (if supplied)
+	if params.Ratio.GT(sdk.ZeroDec()) {
+		denoms := k.GetCollateralTypes(ctx)
+		for _, denom := range denoms {
+			ratio, err := k.CalculateCollateralizationRatioFromAbsoluteRatio(ctx, denom, params.Ratio, "liquidation")
+			if err != nil {
+				continue
+			}
+			cdpsUnderRatio := k.GetAllCdpsByCollateralTypeAndRatio(ctx, denom, ratio)
+			matchRatio = append(matchRatio, cdpsUnderRatio...)
+		}
+	}
+
+	var commonCDPs types.CDPs
+	// If no params specified, fetch all CDPs
+	if len(params.CollateralType) == 0 && len(params.Owner) == 0 && params.ID == 0 && params.Ratio.Equal(sdk.ZeroDec()) {
+		commonCDPs = k.GetAllCdps(ctx)
+	}
+
+	// Find the intersection of any matched CDPs
+	if len(matchCollateralType) > 0 {
+		commonCDPs = matchCollateralType
+	}
+	if len(matchOwner) > 0 {
+		if len(commonCDPs) > 0 {
+			commonCDPs = FindIntersection(commonCDPs, matchOwner)
+		} else {
+			commonCDPs = matchOwner
+		}
+	}
+	if len(matchID) > 0 {
+		if len(commonCDPs) > 0 {
+			commonCDPs = FindIntersection(commonCDPs, matchID)
+		} else {
+			commonCDPs = matchID
+		}
+	}
+	if len(matchRatio) > 0 {
+		if len(commonCDPs) > 0 {
+			commonCDPs = FindIntersection(commonCDPs, matchRatio)
+		} else {
+			commonCDPs = matchRatio
+		}
+	}
```

Comment:
This logic produces slightly confusing results like:

`kvcli q cdp cdps --collateral-type nonsense --ratio 2.0 ` returns all 2.0 ratio cdps, rather than no cdps.
`kvcli q cdp cdps --owner <address> --collateral-type btc` returns all of a user's cdps when they don't have a btc cdp, rather than no cdps

I think whether the `matchX` results are included in the final intersection should be decided by whether the query param was 0 (or "" etc), rather than whether the length of results was 0.

Reply:
Agreed, updated in https://github.com/Kava-Labs/kava/pull/644/commits/5a1f7e84a928fffd729cd2fa9716d05fe1c319ae

Reply:
Looks like the second example above still produces the same results

Reply:
Addresed in https://github.com/Kava-Labs/kava/pull/644/commits/d01f9b8e81dfa3be585543e23428d77ba1485e66